hw_diag/utilities/quectel.py

```diff
@@ -150,121 +168,151 @@ def _do_upgrade(desired_fw_version: str) -> bool:
             return False
+        return True
+    except Exception as e:
+        logging.info(f"failed to update setting {setter} to {desired_value}")
+        logging.error(f"error: {e}")
+        return False
 
-        old_value = getter_method()
-        if old_value == desired_value:
-            logging.info(f"Setting is already correct {old_value}")
+
+def update_setting_with_rollback(getter: str, setter: str, desired_value: str) -> bool:
+    try:
+        modem = find_eg25g_modem()
+        if not modem:
+            logging.debug("EG25G modem not found")
             return False
 
-        internet_before_mode_change = is_internet_accessible()
+        update_needed, old_value = setting_needs_update(modem, getter, setter, desired_value)
+        if not update_needed:
+            return False
 
-        logging.info(f'actual value : {old_value} needs to be: {desired_value}')
-        setter_method(desired_value)
-        new_value = getter_method()
-        if (old_value == new_value):
-            logging.error("the value didn't set correctly, will be tried again")
+        if not update_setting(modem, getter, setter, desired_value):
             return False
 
+        internet_before_mode_change = is_internet_accessible()
+
         # we have verified the setting was set correctly, lets reset.
         reset_modem_and_modem_manager(modem)
 
         internet_after_mode_change = is_internet_accessible()
 
         # we had internet before but lost it, restore original value
-        if (internet_before_mode_change and not internet_after_mode_change):
-            logging.error(f'internet lost after setting {desired_value}')
+        if internet_before_mode_change and not internet_after_mode_change:
+            logging.error(f"internet lost after setting {desired_value}")
             modem = find_eg25g_modem()
             if not modem:
-                logging.info('EG25G modem not found, setting old value will be skipped')
+                logging.info("EG25G modem not found, setting old value will be skipped")
                 return False
             increment_retry_count(setter)
-            logging.info(f"restting to {old_value}")
-            setter_method(modem, old_value)
+            logging.info(f"resetting to {old_value}")
+            update_setting(modem, getter, setter, old_value)
     except dbus.exceptions.DBusException as e:
-        logging.error(f'failed to set AT over dbus with error: {e.get_dbus_message()}')
+        logging.error(f"failed to set AT over dbus with error: {e.get_dbus_message()}")
         return False
     except Exception as e:
-        logging.error(f'failed to set AT Setting: {e}')
+        logging.error(f"failed to set AT Setting: {e}")
         return False
     return True
 
 
 def firmware_upgrade_with_rollback() -> bool:
     modem = find_eg25g_modem()
     if not modem:
-        logging.debug('EG25G modem not found')
+        logging.debug("EG25G modem not found")
```

Comment:
Does it even make sense to run the AT commands before this? I'd definitely wait to download until here.

Reply:
AT commands as well as download checks for modem on their own. 
Reason: we start and stop modem multiple times to check for connectivity impact, modem might not be ready for dbus queries or might have changed its dbus object path.